tactics/rewrite.ml

```diff
@@ -1338,49 +1340,70 @@ module Strategies =
       let rec aux input = (f { strategy = fun input -> check_interrupt aux input }).strategy input in
       { strategy = aux }
 
+    let fix_tac (f : 'a pure_strategy -> 'a pure_strategy Proofview.tactic) : 'a pure_strategy Proofview.tactic =
+      let forward_def = ref (fun _ -> assert false) in
+      f {strategy = fun input -> check_interrupt !forward_def input} >>= fun f ->
+      forward_def := f.strategy;
+      Proofview.tclUNIT f
+
+    let all_subterms (s : strategy) : strategy = subterm true default_flags s
+
+    let one_subterm (s : strategy) : strategy = subterm false default_flags s
+
     let any (s : 'a pure_strategy) : 'a pure_strategy =
       fix (fun any -> try_ (seq s any))
 
     let repeat (s : 'a pure_strategy) : 'a pure_strategy =
       seq s (any s)
 
-    let bu (s : 'a pure_strategy) : 'a pure_strategy =
```

Comment:
Why this specialization, and the other ones below? The more general type makes it clearer by parametricity that it doesn't need to access the state.

Reply:
It's not a specialization, the type was already `unit pure_strategy` (`= strategy`), despite the signature. For whatever reason, OCaml treats this `'a` as unifiable, and in this case it instantiated it with `unit` due to some dependency.

Reply:
Okay, I've just looked at it, and this is because initially `all_subterms` and `one_subterm` were defined as values instead of a functions (i.e. `let all_subterms = subterm true default_flags`), so the monomorphism restriction kicked in and forced OCaml to unify with the first thing it had by its hand. Tbh I find this behaviour quite misleading.

Anyway, I'm fixing it

Reply:
Fixed in 8ce7a32148. Note that `one_lemma`, `lemmas`, `hints` and `old_hints` remain specialized because they rely on `apply_lemma` which explicitly forces `state=()`.